File: 152_MaximumProductSubarray/MaximumProductSubarray.cxx
#include<vector>

using namespace std;

class Solution {
public:
    int maxProduct(vector<int>& nums) {
        if(nums.size() == 0) return 0;
        int maxF = nums[0], minF = nums[0], ans = nums[0];
        for (int i = 1; i < nums.size(); ++i) {
            if(nums[i] < 0) swap(maxF, minF);
            maxF = max(maxF * nums[i], nums[i]);
            minF = min(minF * nums[i], nums[i]);
            ans = max(maxF, ans);
        }
        return ans;
    }
};

File: 0152_MaximumProductSubarray/MaximumProductSubarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if not nums:
            return 0
        MAX = nums[0]
        MIN = nums[0]
        ans = nums[0]
        for n in nums[1:]:
            if n < 0:
                MAX, MIN = MIN, MAX
            MAX = max(n*MAX, n)
            MIN = min(n*MIN, n)
            ans = max(MAX, ans)
        return ans